Userland/Libraries/LibTest/TestCase.h

```diff
@@ -27,10 +47,71 @@ class TestCase : public RefCounted<TestCase> {
     {
     }
 
+    TestCase(DeprecatedString const& name, TestFunction& fn, bool is_benchmark)
+        : m_name(name)
+        , m_function(move(fn))
+        , m_is_benchmark(is_benchmark)
+    {
+    }
+
     bool is_benchmark() const { return m_is_benchmark; }
     DeprecatedString const& name() const { return m_name; }
     TestFunction const& func() const { return m_function; }
 
+    static NonnullRefPtr<TestCase> randomized(DeprecatedString const& name, TestFunction&& test_function)
+    {
+        using namespace Randomized;
+        TestFunction test_case_function = [test_function = move(test_function)]() {
+            for (u32 i = 0; i < MAX_GENERATED_VALUES_PER_TEST; ++i) {
+                bool generated_successfully = false;
+                u8 gen_attempt;
+                for (gen_attempt = 0; gen_attempt < MAX_GEN_ATTEMPTS_PER_VALUE && !generated_successfully; ++gen_attempt) {
+                    // We're going to run the test function many times, so let's turn off the reporting until we finish.
+                    disable_reporting();
+
+                    set_current_test_result(TestResult::NotRun);
+                    run_with_randomness_source(RandomnessSource::live(), test_function);
+                    switch (current_test_result()) {
+                    case TestResult::NotRun:
+                        VERIFY_NOT_REACHED();
+                        break;
+                    case TestResult::Passed: {
+                        generated_successfully = true;
+                        break;
+                    }
+                    case TestResult::Failed: {
+                        generated_successfully = true;
+                        RandomRun first_failure = randomness_source().run();
+                        RandomRun best_failure = shrink(first_failure, test_function);
+
+                        // Run one last time with reporting on, so that the user can see the minimal failure
+                        enable_reporting();
+                        run_with_randomness_source(RandomnessSource::recorded(best_failure), test_function);
+                        return;
+                    }
+                    case TestResult::Rejected:
+                        break;
+                    case TestResult::Overrun:
+                        break;
+                    default:
+                        VERIFY_NOT_REACHED();
+                        break;
+                    }
+                }
+                if (!generated_successfully) {
+                    // The loop above got to the full MAX_GEN_ATTEMPTS_PER_VALUE and gave up.
+                    // Run one last time with reporting on, so that the user gets the REJECTED message.
+                    enable_reporting();
```

Comment:
should we not alway re-enable reporting? What if the next test case is not a randomized test case? I think this call should go outside of this if statement.